### DISCUSSION TITLE
Feedback for Google Analytics import

### DISCUSSION THREAD

**Discussion by ukutaht** [plausible/analytics#1794] [Anything Else]
There is a bigger discussion for general import/export with all tools: https://github.com/plausible/analytics/discussions/211

Since mostly people are coming from GA, we thought that offering a smooth experience for that is the most important. We have been working on it since last year and it's finally ready to be tested on https://plausible.io!

We are looking for some early adopters to give it a go and leave feedback down below about how the feature feels. Mainly we're trying to iron out any issues/bugs/errors with the process and figure out the edge cases, of which we have already found many.

You can start the import process from your site settings > general > import from Google Analytics. [Here are the relevant docs](https://plausible.io/docs/google-analytics-import) and there are some instructions as you go through the process. Please let us know if anything is unclear.

Looking forward to all the feedback!

**Comment by Vigasaurus** [comment]
That was honestly surprisingly flawless (and fast for importing 5 years of data!). Some numbers are a bit confusing to me, namely the unique visitors discrepancy in the full 5-year view - but the rest all seems perfect. And of course the visit duration etc for the single-day aren't expected to work given the resolution of the GA data (but maybe that's a bug and not the intended behavior?), but even the rest of the single-day stats seem pretty perfect. 

I haven't gotten to testing this on my self-hosted instance yet, nor deep diving into the page-specific results/referrers etc, but I will do that soon (these results + screenshots are from prod). 

-- 5 Years of Stats -- 

![image](https://user-images.githubusercontent.com/19434157/160098338-03cb56be-013d-4923-8f8c-508e3a7199c7.png)
![image](https://user-images.githubusercontent.com/19434157/160098352-fd38fa3c-20f2-4a11-9bb1-3fea4e6c16e5.png)

-- Single Day --

![image](https://user-images.githubusercontent.com/19434157/160098369-cb981427-5f3d-40e8-87db-5dc77d6e3969.png)
![image](https://user-images.githubusercontent.com/19434157/160098382-545c642f-81e3-4f32-8513-49b557a6e1d7.png)

**Reply by ukutaht** [reply]
> That was honestly surprisingly flawless

That's great to hear!

The discrepancy in unique visitors is because we only import day-level aggregated data. To get the 5-year number we just sum up the daily visitor numbers instead of doing a proper unique calculation like GA is doing. Other metrics are really easily composable like this but unique visitors is something that doesn't work well when dealing with pre-aggregated data.

The same limitation is why you don't have hourly data in the dashboard. It's impossible to get the raw event-level data from GA so we have to use some aggregates. We could get closer to GA native experience by importing hourly, daily, monthly and yearly data all in separate tables and going to the correct table based on your date range in Plausible.

For the sake of simplicity, we decided to just import day-level aggregates and accept some limitations (like not having hourly data or proper uniqueness across multiple days). It's so much easier to get it right this way, plus it saves on data storage.

> And of course the visit duration etc for the single-day aren't expected to work given the resolution of the GA data (but maybe that's a bug and not the intended behavior?)

It should work. I'm curious what went wrong there. Could you replicate the exact API request that we use to get this data to see if the API is returning bad data or whether we're not handling it well? You can do it really easily by going to [this url](https://ga-dev-tools.web.app/query-explorer/) and replicating the request like this:

![Screenshot from 2022-03-25 14-52-24](https://user-images.githubusercontent.com/3731516/160124338-64b4e0ed-1a03-43c5-bfd6-97f3f729ce1c.png)

**Reply by Vigasaurus** [reply]
Looking back at it, it seemingly is related to the session duration being returned in scientific form (there are more days that are showing a zero-duration, but most of them are actually fine - turns out I just picked the period of time with the most overall session time). All the days that are showing zero-length are returning in this format. 

Honestly this is the stupidest thing I've ever seen in an API - it seems so strange for the number to conditionally change its number format. 

![image](https://user-images.githubusercontent.com/19434157/160136461-95df5994-e9dd-4969-a2fb-8b8f880cf3f9.png)

**Reply by ukutaht** [reply]
@Vigasaurus Amazing, thank you so much for helping with the debugging! I'll make sure we can parse scientific notation as well.

> Honestly this is the stupidest thing I've ever seen in an API - it seems so strange for the number to conditionally change its number format.

Hmm maybe it's because they're worried about overflowing `Int` types or something? Very unexpected in any case :)

**Reply by ukutaht** [reply]
Could you please clear your imported data and try again? Scientific notation should be parsed now.

**Reply by Vigasaurus** [reply]
@ukutaht Looks good now!
![image](https://user-images.githubusercontent.com/19434157/160916161-d30672b9-2cc8-4753-aede-70fb5cc7c819.png)

**Comment by joffreypersia** [comment]
Thank you so much. It is working like a charm ! 

I am impressed. 

I will switch all my websites to plausible and pay for the monthly sub !

**Comment by metmarkosaric** [comment]
thank you! the docs have now been published at https://plausible.io/docs/google-analytics-import

there are some known limitations to how the imported data behaves compared to the native data and they are listed in the docs

and we'll change the link in the app to point to the correct docs link

**Comment by DieterHolvoet** [comment]
Is there any risk in testing this now? I have really been looking forward to this feature, but I wouldn't want to risk corrupting my existing data.

**Reply by metmarkosaric** [reply]
no risks from what we know. you can always delete your imported data with one click in the settings and it doesn't affect your Plausible data

**Comment by MSchmidt** [comment]
I've ran into an error right after connecting to my Google Account. I've submitted the report to Sentry.

**Reply by ukutaht** [reply]
Oh no!

Looks like the GA API response is missing the `websiteUrl` field which we attempt to display. I've added a fallback so you should see a numeric ID instead in the dropdown.

Could you try again? I don't really understand why this field isn't being returned and I cannot find anything from Google's docs. But I've added some more diagnostics so when you go through the flow I should see the full API response.

**Reply by MSchmidt** [reply]
Worked great now. With so much legacy data inside plausibe now, the datepicker and graph could need some work now though.

**Reply by ukutaht** [reply]
Thanks!

Besides adding and 'All time' option which I'll do soon, what other suggestions do you have?

**Reply by bjarte** [reply]
> With so much legacy data inside plausibe now, the datepicker and graph could need some work now though.

I agree. I hope in the future we can have longer timespans than 12 months in the graph.

It is possible to use "Custom range" to select a previous year, but it is not very user friendly yet. Takes a lot of clicks, and after you have done your selection, the selected year is not displayed anywhere, only the selected dates.

**Reply by ukutaht** [reply]
>  the selected year is not displayed anywhere, only the selected dates.

Ah good point, this has tripped me up in the past as well. Adding the year would be very useful. I'll have to see how it fits in the UI.

**Reply by MSchmidt** [reply]
Some quick ideas in no particular order
1) show year numbers when more than 1 year is relevant to the current time span
2) allow moving through timeframe. E.g. when 12 months are shown it would be nice to have arrows to move 12 month back/forth with one click.
3) support manual entering in custom range selector
4) whole separate feature: compare time spans

**Reply by ukutaht** [reply]
Thanks @MSchmidt!

1. Agreed.
2. Agreed.
3. Haven't considered this. I've always thought the best UI for selecting a custom range would be to pick it straight from the chart itself by dragging your mouse. This is sort of close to what I've had in mind: https://dribbble.com/shots/5263343--Date-Range-Selector. This isn't mutually exclusive with having date inputs though.
4. This is in the pipeline. Coming soon :)

**Reply by MSchmidt** [reply]
Nice. Having something better than a manual input would be great. Sure. But the current issue is from getting zooming out/getting to a multi-year-timeframe. So the mouse-handle would work if there was a quick way to go to "whole range of all data" first and then use the mouse to "zoom in" to the right spot. Would work I guess if technical plausible (no pun intended).

**Reply by bjarte** [reply]
Another feature I'm missing now that I have many years of data in Plausible, is the possibility to compare two timespans. Compare this year with last year, for example.

**Comment by bjarte** [comment]
The import tool seems to work really well. 

I had one little hiccup during import: 

I have two websites with GA data. One with lots of traffic and one with almost none. Both have GA data from 2016 to today.

I started both imports straight after each other, the big one first. When the second site was completed (which was almost immeditialy), I noticed the first import had failed.

I re-ran the import for the big site, and after a little while, I got an e-mail that it was complete.

I'm not sure why the first import failed, but the error messages told me to try again, which I did, and then it worked.

**Reply by ukutaht** [reply]
Thanks! I've added a retry mechanism to the API calls so hopefully the import will be more robust in the future. I'll also look at options to retry the whole job if it fails with an API timeout (which is the main error I'm seeing)

**Comment by MSchmidt** [comment]
The whole process is really nice and simple. It would be interesting to know what "Enable Users Metric in Reporting" actually does and why it needs to be off.

I was very positively surprised by the automatic time range selection that was done for me. I was totally expecting double data from the moment I enabled plausable. But you were smarter and handled it well. Awesome.

**Reply by ukutaht** [reply]
I don't know exactly what difference "Enable Users Metric in Reporting" makes but if it's enabled the API just returns 0 visitors for each day. Not very useful.

It's only necessary for the duration of the import, you can switch the setting back after the import is done.

**Comment by MSchmidt** [comment]
Not strictly an import issue but the import showed a pretty stark difference in visitors for me. I've switchen to Plausible on Jan 8 with this particular site and this is what it looks like:

<img width="1091" alt="Screenshot 2022-04-01 at 09 20 22" src="https://user-images.githubusercontent.com/38004/161214874-27d60ee9-416b-4a70-b5f5-f732c9e67d60.png">

Any idea why this might be? I expected an increase since no opt in is needed for plausible.

**Reply by metmarkosaric** [reply]
hi @MSchmidt! very weird for sure as most sites see higher traffic with Plausible than with GA. couple of ways to figure out what's going on: can you confirm that you imported the stats of the same website as the one you're tracking in Plausible? can you check whether Plausible is correctly installed on all of the pages that GA was installed on? feel free to [contact our customer support](https://plausible.io/contact) and we can help you troubleshoot your integration. thanks!

**Reply by MSchmidt** [reply]
I'm pretty sure it's the correct site. And I should have seen the difference earlier. The Import just surfaced this. Will go the customer support route for this one since I guess it's really not an import topic. Sorry for that.

**Reply by metmarkosaric** [reply]
np! we'll take a look at your site to see what's going on and we'll get back to you asap

**Reply by MSchmidt** [reply]
@metmarkosaric So your team discovered that the difference comes from bot traffic in my legacy GA data. There is no way to filter this after the fact. Enabling the filter on GA only affects newly collected data. Which of course is already done by Plausible. So the question is whether you can apply your internal filtering logic to imports as well to some degree.

**Reply by metmarkosaric** [reply]
thanks @MSchmidt! we import data in aggregate only so unfortunately we cannot apply our bot filtering on the GA traffic. it's best to remove any bot data and other data you don't want imported directly from GA before making the import into Plausible. there should be a way in GA to exclude bot traffic

**Reply by MSchmidt** [reply]
I've tried two methods of excluding bot data from GA and both only apply to future traffic. I didn't find a way to filter the existing dataset yet :(

**Reply by metmarkosaric** [reply]
ah that's too bad

**Reply by ukutaht** [reply]
Theoretically we could add some sort of best-effort bot filtering to our import. It wouldn't be perfect and it wouldn't align exactly with how we do it for Plausible native data but it would be something.

I remember in your case @MSchmidt there bot traffic is clearly coming from `Browser = StatusCake` or something like this. When importing data, we could theoretically apply a filter to only fetch GA stats where `Browser != StatusCake`.

Enumerating all possible bot user-agents in GA is quite difficult and probably impossible to be comprehensive. I don't know if it's worth attempting to filter out bots during import...

**Comment by metmarkosaric** [comment]
A new version of our GA import tool is now live on the site and this one should work fine even for very large sites. If the import failed on you before, could you please try it again and let us know if it works now? Thank you!

**Reply by MSchmidt** [reply]
I've imported about 8.5M records successfully. Was still pretty quick.

**Reply by metmarkosaric** [reply]
great to hear! thanks for the update!

**Comment by metmarkosaric** [comment]
Universal Analytics, the current version of Google Analytics, [will be sunset](https://plausible.io/blog/universal-google-analytics-is-dead) and will stop counting stats on July 1st 2023. There’s no way to import your historical Universal Analytics stats into the new Google Analytics 4, so we've put a lot of effort into making a Google Analytics stats import tool ourselves. 

You can now import your historical Google Analytics (Universal Analytics) stats into your Plausible dashboard. You can import the stats from your first Google Analytics visitor until your first Plausible Analytics visitor. [See more here](https://plausible.io/docs/google-analytics-import).

**Comment by mrviste** [comment]
Thank you for easy GA import! 
Waiting for the possibility to select historical data in the "Custom range". For now, I can only select within the period of my Plausible tracking...

<img width="1124" alt="Skjermbilde 2022-04-12 kl  18 50 18" src="https://user-images.githubusercontent.com/17169225/163013851-66472fd3-6fe8-4611-9a96-0b26e3095e2c.png">

**Reply by metmarkosaric** [reply]
thanks for the feedback @mrviste! this should be working all fine, i've checked on my sites and the custom range shows the historical time periods too. what website is this? could you please [email our customer support](https://plausible.io/contact) and we'll take a look for you

**Comment by Lucent** [comment]
I started an import yesterday which is still running of data from 14 Nov 2005 to 15 Sep 2020.

I'd also like the option to import all data, as I never ran Analytics concurrently with Plausible and want to combine both as the SPA takes considerable time to expire cache, so Analytics visits are still coming in.

**Reply by metmarkosaric** [reply]
thanks for the feedback @Lucent! the import usually takes only a few minutes. please cancel it and try again. we found out that for some accounts, especially those with historical data from before August 2016, changing the user metric settings is necessary to get the data out of the API so please do make sure to follow that extra step.

**Reply by ukutaht** [reply]
> I'd also like the option to import all data

By this do you mean that you would like to change the end date of the import date range?

**Reply by Lucent** [reply]
Yes, setting an end date would be the simplest way of doing it.

My second import completed, but only shows data from Sep 2009 to Nov 2012. "Enable Users Metric in Reporting" was always disabled.

One special request that'll help me transition entirely from Analytics: there was a period from Sept 2020 to June 2021 when I was concurrently testing Plausible and Google Analytics. Can you scrub the data Plausible recorded in this time period (inclusive) from my account so that the Analytics import can overwrite it?

**Reply by metmarkosaric** [reply]
Google Analytics API sometimes randomly returns empty data. It's intermittent and random. Trying again multiple times will eventually return what you need. We'll try to build automatic retries in our GA import tool when we see empty data as this has happened to other people as well.

We don't recommended you to remove native data to replace it with imported data as the native data is superior to the imported data but it is possible if you wish to do so. You can [reset the Plausible stats](https://plausible.io/docs/reset-site-data) first and make the import after that.

**Reply by Lucent** [reply]
There's no way to scrub out a range of data? I fully transitioned to Plausible from Analytics back in January and don't want to lose any data since then.

**Reply by Lucent** [reply]
I've attempted a dozen imports of that 15-year span without success. I see it now correctly issues a failure if it doesn't complete rather than show a partial import.

**Comment by tonysaad** [comment]
![Screenshot 2022-06-02 at 12 47 01](https://user-images.githubusercontent.com/1137569/171613429-e659e3c5-c077-4137-bcdc-7794651bf950.png)
I've been trying for the last few weeks to import GA data, but I can never seem to find my view. it's always an empty dropdown.
What might be wrong ? any ideas ?
thanks

**Reply by metmarkosaric** [reply]
hi @tonysaad! this typically happens if you don't have any valid UA properties in the google account that you integrated with. google basically sends us an empty list. perhaps you have ga4 properties only? or you don't have the rights to any properties?

**Comment by tonysaad** [comment]
That’s true it’s only ga4, is that not supported ?

**Reply by metmarkosaric** [reply]
yeah, we only support UA at the moment. GA4 is not widely used yet so we built the import tool to help those that didn't want to move to GA4. we may be able to build a GA4 import tool in the future too when that becomes the default GA at some point next year

**Reply by tonysaad** [reply]
Thanks, Yeah our usecase was different then

**Comment by boazsender** [comment]
I've been testing a self hosted plausible instance to replace my google analytics setup, and am really enjoying the product. Thanks so much! I don't see the google analytics import available in the self-hosted version. Is there a plan for bringing this over to the self-hosted at any point, and is there a way to use this feature self-hosted in the meantime? Thanks again for all your amazing work on this. I truly appreciate having such a solid alternative to turn to.

**Reply by metmarkosaric** [reply]
Thanks @boazsender! Our self-hosted version is not updated as frequently as the Plausible managed hosting in the cloud, which changes daily. This means that the latest features won't be immediately available. Think of self-hosted as a long-term release. We stack all the new features and fixes that have been battle-tested on the hosted version and release a new self-hosted version twice a year or so. We don't have any more specific timelines than this as we have limited resources, and the self-hosted release depends on the free capacity of our developers.

**Reply by boazsender** [reply]
Thanks so much @metmarkosaric, I'll watch the release page.

**Comment by maximilianschmid** [comment]
Where is the "import from google analytics" button as stated in the documentation under point two here: https://plausible.io/docs/google-analytics-import#how-to-import-your-historical-stats-from-google-analytics-to-plausible-analytics 

![missing google analyics import](https://user-images.githubusercontent.com/574203/180961039-af5371e8-c552-4981-8574-316cf7cf6dbe.png)

I do have a subscription active, but the Import feature still does not show up. Any suggestions?

**Reply by metmarkosaric** [reply]
Hi @maximilianschmid! We had such high demand for Google Analytics imports that it negatively affected the stability of the rest of our services so we had to disable the functionality temporarily. There's a note about this just above the docs link you shared. So we're currently working on some tweaks and we're hoping to return the import functionality as soon as possible. Our developers are currently testing a new implementation so it should not be too long. See https://github.com/plausible/analytics/pull/2046. Thanks for your patience and do stay tuned for an update!

**Reply by maximilianschmid** [reply]
Thank you for this info. If you want important notes to be seen, maybe you should make them look something like this:

![CleanShot 2022-07-28 at 16 07 52@2x](https://user-images.githubusercontent.com/574203/181525957-facec1e4-d608-4b7f-9eb2-6a4157c6183e.png)

…or just place a hint in the app itself where the GA import functionality would be.

**Reply by metmarkosaric** [reply]
thanks! fixed it now in the docs

**Comment by futureshocked** [comment]
Hello @metmarkosaric , do you have an update on the Google Analytics import function? 

I'm keen to migrate everything to Plausible, but I need my GA history.

**Reply by metmarkosaric** [reply]
hi @futureshocked! one of our developers just [submitted a PR](https://github.com/plausible/analytics/pull/2184) for the last remaining issue that we have encountered in our extensive testing so it's almost ready to go. should be out very very soon

**Reply by futureshocked** [reply]
Thank you, @metmarkosaric , looking forward to it.

**Reply by metmarkosaric** [reply]
Google Analytics import is now back!

https://plausible.io/docs/google-analytics-import

**Reply by futureshocked** [reply]
Hi @metmarkosaric , that's awesome, thank you!

I imported history for one of my sites, but the other shows "no data found", even though there's plenty of data.

<img width="467" alt="2022-09-13_18-10-53" src="https://user-images.githubusercontent.com/25273/189847699-1bc2847d-8980-483b-a481-34fb5730c1d2.png">

Maybe a bug? something else I can try?

**Reply by metmarkosaric** [reply]
hmm weird! thanks for reporting @futureshocked! 

is there data in there from before your joined Plausible? we only import up to the day you started counting stats with Plausible. 

or do you have any strict data retention limits in place in your GA settings that for instance removes stats after 16 months? you can look at  "Tracking Info" section in GA and click on "Data Retention". 

if it's neither of these two then please do contact our customer support so we can take a closer look https://plausible.io/contact

**Reply by futureshocked** [reply]
Hi @metmarkosaric , 

I'm sorry, I had selected the wrong source.

However, the correct source (with retention of 26 months), only imported a segment of the data (pls see below). 

<img width="804" alt="2022-09-13_18-59-17" src="https://user-images.githubusercontent.com/25273/189860363-fc116e9f-abf5-429c-a32e-f9d28d9c3961.png">

However, the source has later data (i.e. from September 2020 to now).

Maybe there is a limit to the import tool?

**Reply by metmarkosaric** [reply]
Cool. Data is imported from your first Google Analytics visitor until your first Plausible Analytics visitor so perhaps you started using Plausible then?

**Comment by futureshocked** [comment]
I think I did a trial back then but did not sign up until now for a paid
account.

Is it possible to reset so I can get the rest of the history imported?

On Tue, 13 Sep 2022 at 19:08, Marko Saric ***@***.***> wrote:

> Cool. Data is imported from your first Google Analytics visitor until your
> first Plausible Analytics visitor so perhaps you started using Plausible
> then?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/plausible/analytics/discussions/1794#discussioncomment-3633404>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAGFONUIPN4KBRIIVYR4NLV6BAA7ANCNFSM5RTS4U2Q>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 

_________________________________________________
Peter Dalmaris, PhD

**Reply by metmarkosaric** [reply]
I see. This isn't recommended as the native data is superior to the imported data, but it's possible if you wish to do so. You can [reset the Plausible stats](https://plausible.io/docs/reset-site-data) first and make the import after that.

**Reply by futureshocked** [reply]
Yes, that worked! 

I only had a week of Plausible data. After the reset, I was able to import all of GA (~3 years). 

Thank you @metmarkosaric !

**Reply by metmarkosaric** [reply]
cool, nice to hear that!

**Comment by czj** [comment]
We have tried import in march and it failed, but as of 2022-10-15 (yesterday) the import worked very well.

We have data from 2005 to 2015 imported correctly but nothing from 2015 to 2020 (when we stopped using UA).

<img width="830" alt="2022-10-15 - 08h31m09" src="https://user-images.githubusercontent.com/14306/195972873-8dac5a41-e6f2-474e-8334-a6f4bfa22756.png">

We followed instruction on [the import guide](https://plausible.io/docs/google-analytics-import).

UA show data from January 2015 :

<img width="1148" alt="2022-10-15 - 08h35m42" src="https://user-images.githubusercontent.com/14306/195973055-4e73f7c1-6abc-4fc3-92ee-95f4322b7773.png">

(end of data from mid 2020 is normal : we switched to GA4)

**Reply by metmarkosaric** [reply]
thanks for reporting @czj! could you please share the url of your site or even better send an email to our customer support so we can take a look https://plausible.io/contact

**Reply by czj** [reply]
Will do, of course :-)
